### PR TITLE
feat: support local IAPKit receipt verification

### PR DESCRIPTION
**Pull request by hyochan**
## Summary

- Add optional `iapkit.baseUrl` support across OpenIAP native packages and framework bridges, with HTTP(S)-origin validation and same-deployment API-key guidance.
- Add `Local (IAPKit)` to the React Native and Expo Martie examples in the exact order `Local (Device)` → `Local (IAPKit)` → `IAPKit` → `None (Skip)`.
- Harden the compiled IAPKit server smoke test, add a dedicated Martie local-receipt E2E workflow/skill, and make Godot Apple framework builds/load checks fail closed.
- Bump the OpenIAP spec patch from 2.3.0 to 2.3.1 and add the API availability note to generated SDK types and public docs.

## Root cause

IAPKit receipt verification was fixed to the hosted endpoint in native and Vega paths, so a device purchase could not be routed through the locally compiled IAPKit server. The examples also had no distinct local-server choice, and the server smoke probe could accidentally succeed against an older process already owning the port.

### Live receipt vertical

- Physical iPhone 13 mini, Expo example `dev.hyo.martie`, Apple sandbox.
- SKU: `dev.hyo.martie.10bulbs`.
- Compiled local `packages/kit` server backed by the real Martie Dev Convex deployment.
- Correlation ID: `891e06ad-01d0-419e-8f2a-c77b156ae037`.
- Local server returned HTTP 200 with `isValid: true` / `ready-to-consume`; the app finished the transaction.
- Matching Dev Convex purchase row: `k9751qztk3dp92y9pq3vks3dcx8ae84s`.
- No additional purchase was made during the final review.

### Device UI preview

- Built and installed the current React Native example on a physical Pixel 2.
- Confirmed all four verification choices appear in the requested order and selecting `Local (IAPKit)` updates the purchase screen.
- A short screen recording is attached in the PR conversation.

### Automated and build checks

- GQL generation and tests: 47 passed.
- IAPKit: lint, Prettier, 51 test files / 572 tests, production build, compiled-server smoke including `POST /v1/purchase/verify → 400`.
- React Native and Expo focused/full adapter and example tests, including Kepler/Vega URL edge cases.
- Apple URL tests, Google validator tests, Flutter tests/analyze, KMP Android/iOS compiles, MAUI net9/net10 builds.
- Godot Android bridge test; iOS/macOS paired frameworks rebuilt; native ClassDB load plus 103 types and 73 wrapper tests passed.
- `bun audit:parity`, `bun audit:docs`, `bun audit:release-state`, docs typecheck/format, and `git diff --check`.

## Release gate

After merge, release the Apple package and required Google flavors first. Then sync the new native version pins and rebuild downstream artifacts before releasing React Native, Expo, Flutter, KMP, or Godot. Monorepo tests use local-project substitution; publishing downstream SDKs against the current Apple/Google pins would otherwise discard `baseUrl`.

## Summary by CodeRabbit

* **New Features**
  * Added Local (IAPKit) purchase verification with a configurable IAPKit server base URL.
  * Added a verification method selector/modal, including Local Device, Local (IAPKit), hosted IAPKit, and Skip.
  * Extended the configurable base URL through React Native, Expo, Flutter, Kotlin, Swift, MAUI, and Godot integrations.
* **Documentation**
  * Expanded setup and end-to-end receipt verification guidance for hosted vs local flows, including device routing and validation steps.
* **Bug Fixes**
  * Improved custom IAPKit endpoint handling with stricter HTTP(S) origin validation and safer request URL construction.
* **Tests**
  * Added/expanded unit, adapter, and E2E coverage for local receipt verification and base URL forwarding.

### CODE REVIEW

**Comment by hyochan** [comment]
Physical Pixel 2 preview (no purchase): verification modes appear in the required order — Local (Device), Local (IAPKit), IAPKit, None (Skip) — and selecting Local (IAPKit) is reflected in the UI.

https://github.com/user-attachments/assets/038338ba-023b-402c-8acb-6ae8c2491d1a

File: .claude/commands/e2e-tests.md [hunk 1]
```diff
@@ -76,6 +80,99 @@ Notes:
...
+prerequisite. A server health check, build, mocked receipt, or HTTP 400 route
+probe does not count as a live receipt PASS.
+
+Always after the row, including `PASS`, `FAIL`, or `BLOCKED`, stop the local
+server and log streams. Remove the reverse rule only when this run created it:
+
+```bash
+if [ "${IAPKIT_REVERSE_CREATED:-0}" = "1" ]; then
+  adb -s "$ANDROID_SERIAL" reverse --remove tcp:3100
+fi
```
[the remark this answers is not in the document]
Comment (hyochan) [reply .claude/commands/e2e-tests.md:178]: Fixed in 38b065aa. Cleanup now re-reads the current reverse rules and removes tcp:3100 only while the exact mapping created by this run is still present; changed or unreadable mappings are left untouched.

File: .claude/commands/e2e-tests.md [hunk 2]
```diff
@@ -76,6 +80,99 @@ Notes:
...
+  IAPKIT_REVERSE_CREATED=0
+elif adb -s "$ANDROID_SERIAL" reverse --no-rebind tcp:3100 tcp:3100; then
+  IAPKIT_REVERSE_CREATED=1
+else
+  IAPKIT_REVERSE_CREATED=0
+  IAPKIT_REVERSE_BLOCKED=1
+  echo 'BLOCKED: could not create tcp:3100 reverse mapping without rebinding' >&2
+fi
+[ "$IAPKIT_REVERSE_BLOCKED" = "0" ]
+```
```
[the remark this answers is not in the document]
Comment (hyochan) [reply .claude/commands/e2e-tests.md:120]: Fixed in 38b065aa. The reverse-rule guard now exits explicitly when adb reverse --no-rebind fails, so the receipt flow cannot continue through a conflicting mapping.

File: libraries/expo-iap/src/vega-adapter.ts [hunk 3]
```diff
@@ -188,6 +188,121 @@ function createVegaError(
   return error;
 }
 
+function isValidIpv4Address(address: string): boolean {
+  const octets = address.split('.');
+  return (
+    octets.length === 4 &&
+    octets.every((octet) => /^\d{1,3}$/.test(octet) && Number(octet) <= 255)
+  );
+}
```
[the remark this answers is not in the document]
Comment (hyochan) [reply libraries/expo-iap/src/vega-adapter.ts:200]: Fixed in 24cc620e. Expo now rejects leading-zero IPv4 octets, including embedded IPv4 in IPv6 literals, before fetch can reinterpret the destination. The targeted adapter suite passes 70/70.

File: packages/kit/scripts/smoke-server.sh [hunk 4]
```diff
@@ -50,23 +51,31 @@ cleanup() {
...
     exit 1
   fi
+  if grep -Fq "IAPKit server listening on :${PORT}" "$LOG_FILE"; then
+    ready=1
+    break
+  fi
   sleep 0.25
 done
 
+if [[ "$ready" -ne 1 ]]; then
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/scripts/smoke-server.sh:72]: Not changing this: smoke-server.sh is explicitly a Bash executable (#!/usr/bin/env bash), and every repository invocation executes ./scripts/smoke-server.sh directly. It already relies on Bash-only BASH_SOURCE, pipefail, local, and multiple existing [[ expressions, so replacing this one condition would neither improve nor promise POSIX sh compatibility. bash -n passes.

File: libraries/expo-iap/example/src/utils/vegaRuntime.ts [hunk 5]
```diff
@@ -46,6 +55,24 @@ function withIapkitEndpoint(
...
+
+  const baseUrl = configuredBaseUrl?.trim();
+  if (!baseUrl) {
+    throw new Error(
+      'EXPO_PUBLIC_IAPKIT_BASE_URL not configured for Local (IAPKit) verification',
+    );
+  }
+
+  return baseUrl;
+}
```
[the remark this answers is not in the document]
Comment (hyochan) [reply libraries/expo-iap/example/src/utils/vegaRuntime.ts:74]: Not changing this: the call occurs only in the else branch after ignore and local have both been excluded, so TypeScript narrows currentVerificationMethod to iapkit-localhost | iapkit. The strict Expo example typecheck passes. Widening the helper to VerificationMethod would hide invalid ignore/local calls and weaken its IAPKit-only contract.

File: libraries/react-native-iap/example/src/utils/vegaRuntime.ts [hunk 6]
```diff
@@ -21,6 +19,24 @@ function withIapkitEndpoint(
...
+
+  const baseUrl = configuredBaseUrl?.trim();
+  if (!baseUrl) {
+    throw new Error(
+      'IAPKIT_BASE_URL not configured for Local (IAPKit) verification',
+    );
+  }
+
+  return baseUrl;
+}
```
[the remark this answers is not in the document]
Comment (hyochan) [reply libraries/react-native-iap/example/src/utils/vegaRuntime.ts:38]: Not changing this: the call occurs only in the else branch after ignore and local have both been excluded, so TypeScript narrows currentVerificationMethod to iapkit-localhost | iapkit. The strict React Native typecheck passes and includes both flows. Widening the helper to string would hide invalid calls and weaken its IAPKit-only contract.

File: packages/kit/scripts/smoke-server.sh [hunk 7]
```diff
@@ -81,6 +90,24 @@ probe() {
...
+  local path="$1"
+  local expected="$2"
+  local code
+  code="$(curl -sS -o /dev/null -w "%{http_code}" \
+    -X POST \
+    -H "Authorization: Bearer openiap-kit_smoke-test-key" \
+    -H "Content-Type: application/json" \
+    --data '{}' \
+    "http://localhost:${PORT}${path}")"
+  if [[ "$code" != "$expected" ]]; then
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/scripts/smoke-server.sh:103]: Not changing this: smoke-server.sh is explicitly a Bash executable (#!/usr/bin/env bash), and every repository invocation executes ./scripts/smoke-server.sh directly. The script already uses Bash-only BASH_SOURCE, pipefail, local, and existing [[ expressions, so changing only this comparison would not make it POSIX sh compatible. bash -n passes.

File: libraries/react-native-iap/src/vega-adapter.ts [hunk 8]
```diff
@@ -150,6 +150,121 @@ function createVegaError(
   return error;
 }
 
+function isValidIpv4Address(address: string): boolean {
+  const octets = address.split('.');
+  return (
+    octets.length === 4 &&
+    octets.every((octet) => /^\d{1,3}$/.test(octet) && Number(octet) <= 255)
+  );
+}
```
[the remark this answers is not in the document]
Comment (hyochan) [reply libraries/react-native-iap/src/vega-adapter.ts:162]: Fixed in 24cc620e. React Native now rejects leading-zero IPv4 octets, including embedded IPv4 in IPv6 literals, before fetch can reinterpret the destination. The targeted adapter suite passes 71/71.

File: .claude/commands/e2e-tests.md [hunk 9]
```diff
@@ -76,6 +80,107 @@ Notes:
...
+IAPKIT_REVERSE_BLOCKED=0
+if printf '%s\n' "$existing_reverse_rules" | \
+  grep -Eq '(^|[[:space:]])tcp:3100[[:space:]]+tcp:3100($|[[:space:]])'; then
+  IAPKIT_REVERSE_CREATED=0
+elif adb -s "$ANDROID_SERIAL" reverse --no-rebind tcp:3100 tcp:3100; then
+  IAPKIT_REVERSE_CREATED=1
+else
+  IAPKIT_REVERSE_CREATED=0
+  IAPKIT_REVERSE_BLOCKED=1
+  echo 'BLOCKED: could not create tcp:3100 reverse mapping without rebinding' >&2
```
[the remark this answers is not in the document]
Comment (hyochan) [reply .claude/commands/e2e-tests.md:115]: Not changing this: Android Platform Tools adb 35.0.2 documents reverse [--no-rebind] REMOTE LOCAL in adb help. The flag is intentional because a mapping can appear after the initial list check; plain reverse could overwrite it. This block instead fails closed when the non-rebinding create cannot proceed.

File: libraries/expo-iap/example/README.md [hunk 10]
```diff
@@ -18,6 +18,25 @@ npm run ios
...
 
+## Purchase Verification
+
+Create the ignored environment file before testing IAPKit:
+
+```bash
+cp .env.example .env
+```
+
+For hosted IAPKit, get a key from the [IAPKit dashboard](https://kit.openiap.dev). Set `EXPO_PUBLIC_IAPKIT_API_KEY` to a key issued by the IAPKit/Convex deployment that the selected server uses. For **Local (IAPKit)**, the key and local server must target the same Convex deployment. Also set `EXPO_PUBLIC_IAPKIT_BASE_URL` to the device-reachable HTTP(S) origin only; do not append `/v1/purchase/verify`. A physical iPhone must use the Mac's LAN address. An Android device connected over USB can use `http://127.0.0.1:3100`: inspect `adb -s "$ANDROID_SERIAL" reverse --list`, reuse an exact `tcp:3100` mapping when present, or create it with `adb -s "$ANDROID_SERIAL" reverse --no-rebind tcp:3100 tcp:3100`. Record whether this run created the rule and remove only that rule during cleanup; if `--no-rebind` fails, use another port instead of overwriting an existing mapping.
```
[the remark this answers is not in the document]
Comment (hyochan) [reply libraries/expo-iap/example/README.md:29]: Not changing this: Android Platform Tools adb 35.0.2 explicitly supports reverse [--no-rebind] REMOTE LOCAL. Keeping --no-rebind prevents the example instructions from overwriting a mapping that appears between inspection and creation.

File: packages/kit/scripts/smoke-server.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/scripts/smoke-server.sh:103]: Not changing this: smoke-server.sh is an explicit Bash executable (#!/usr/bin/env bash), all repository call sites execute it directly, and the script already depends on Bash-only features. Replacing this one [[ expression would not provide POSIX sh compatibility; bash -n passes.

File: packages/kit/scripts/smoke-server.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/scripts/smoke-server.sh:72]: Not changing this: smoke-server.sh is an explicit Bash executable (#!/usr/bin/env bash), all repository call sites execute it directly, and the script already depends on BASH_SOURCE, pipefail, local, and existing [[ expressions. Replacing one comparison would not make it POSIX sh compatible; bash -n passes.

File: .codex/skills/iapkit-e2e-martie/SKILL.md [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+  IAPKIT_REVERSE_BLOCKED=0
+  if printf '%s\n' "$existing_reverse_rules" | \
+    grep -Eq '(^|[[:space:]])tcp:3100[[:space:]]+tcp:3100($|[[:space:]])'; then
+    IAPKIT_REVERSE_CREATED=0
+  elif adb -s "$ANDROID_SERIAL" reverse --no-rebind tcp:3100 tcp:3100; then
+    IAPKIT_REVERSE_CREATED=1
+  else
+    IAPKIT_REVERSE_CREATED=0
+    IAPKIT_REVERSE_BLOCKED=1
+    echo 'BLOCKED: could not create tcp:3100 reverse mapping without rebinding' >&2
```
[the remark this answers is not in the document]
Comment (hyochan) [reply .codex/skills/iapkit-e2e-martie/SKILL.md:131]: Not changing this: Android Platform Tools adb 35.0.2 documents reverse [--no-rebind] REMOTE LOCAL in adb help. The flag preserves the runbook's ownership rule across the list/bind race, and the block explicitly stops as BLOCKED if creation fails.

File: libraries/react-native-iap/example/README.md [hunk 12]
```diff
@@ -29,6 +29,25 @@ This project is part of a **Yarn workspace** structure. All commands should be r
...
 
+### Purchase Verification
+
+Create the ignored environment file from the example before testing IAPKit:
+
+```sh
+cp example/.env.example example/.env
+```
+
+For hosted IAPKit, get a key from the [IAPKit dashboard](https://kit.openiap.dev). Set `IAPKIT_API_KEY` to a key issued by the IAPKit/Convex deployment that the selected server uses. For **Local (IAPKit)**, the key and local server must target the same Convex deployment. Also set `IAPKIT_BASE_URL` to the device-reachable HTTP(S) origin only; do not append `/v1/purchase/verify`. A physical iPhone must use the Mac's LAN address. An Android device connected over USB can use `http://127.0.0.1:3100`: inspect `adb -s "$ANDROID_SERIAL" reverse --list`, reuse an exact `tcp:3100` mapping when present, or create it with `adb -s "$ANDROID_SERIAL" reverse --no-rebind tcp:3100 tcp:3100`. Record whether this run created the rule and remove only that rule during cleanup; if `--no-rebind` fails, use another port instead of overwriting an existing mapping.
```
[the remark this answers is not in the document]
Comment (hyochan) [reply libraries/react-native-iap/example/README.md:40]: Not changing this: Android Platform Tools adb 35.0.2 explicitly supports reverse [--no-rebind] REMOTE LOCAL. Keeping --no-rebind prevents the example instructions from overwriting a mapping that appears between inspection and creation.

File: libraries/maui-iap/src/OpenIap.Maui/Types.cs [hunk 13]
```diff
@@ -4581,6 +4581,12 @@ public sealed record RequestVerifyPurchaseWithIapkitProps
     /// <summary>API key used for the Authorization header (Bearer {apiKey}).</summary>
     [JsonPropertyName("apiKey")]
     public string? ApiKey { get; init; }
+    /// <summary>Available in OpenIAP Spec 2.3.1 / openiap-apple 2.4.0 / openiap-google 2.4.0.</summary>
+    /// <summary>Base URL for the IAPKit server. Defaults to https://kit.openiap.dev.</summary>
+    /// <summary>Set this to a reachable HTTP(S) origin when self-hosting or testing a local IAPKit server.</summary>
+    /// <summary>The apiKey must be issued by the same IAPKit/Convex deployment as this server.</summary>
+    [JsonPropertyName("baseUrl")]
+    public string? BaseUrl { get; init; }
```
[the remark this answers is not in the document]
Comment (hyochan) [reply libraries/maui-iap/src/OpenIap.Maui/Types.cs:4589]: Not changing this generated file: the C# generator intentionally emits one inline summary element per GraphQL description line, and the same established pattern appears throughout Types.cs. The MAUI net9/net10 builds run with GenerateDocumentationFile and TreatWarningsAsErrors and pass without XML-doc warnings. A repository-wide generator formatting redesign is separate from this API addition, and generated copies must not be edited directly.
